src/charm.py

```diff
@@ -440,10 +452,10 @@ def _get_istio_gateway(self):
     def main(self, _) -> None:
         """Perform all required actions the Charm."""
         try:
-            self._check_container_connection()
             self._check_leader()
             self._deploy_k8s_resources()
-            self._update_layer()
+            if self._is_container_ready():
+                self._update_layer()
```

Comment:
I don't think this works as intended.  If `is_container_ready()` returns False, we (correctly, I think) avoid calling `_update_layer()`, but then we continue through the rest of this method (resulting in setting `ActiveStatus` a few lines down.  Previously, this possibility was avoided because `is_container_ready` would have raised an `ErrorWithStatus` that was caught and handled, forcing main the `return` without setting active.  

I personally like the previous way better, but this way can work too so long as it is handled in a way that doesn't set the charm to active

Reply:
A pattern that might work better here is:
```python
def main():
  try: 
    self._check_leader()
    ...
  except ...:
    ...
  self._set_charm_status()
```

where `_set_charm_status` is in charge of looking at the current state (k8s objects, pebble container, etc) and deciding if the charm is up.

That same `_set_charm_status` would be useful in an `update_status` hook, too

Reply:
How about?
```
@@ -454,8 +442,9 @@ class SeldonCoreOperator(CharmBase):
         try:
             self._check_leader()
             self._deploy_k8s_resources()
-            if self._is_container_ready():
-                self._update_layer()
+            if not self._is_container_ready():
+                return
+            self._update_layer()
         except ErrorWithStatus as error:
             self.model.unit.status = error.status
             return
```

Reply:
This way status is not set, because container is still coming up.